### DISCUSSION TITLE
Load the Google AdSense external script to my remix site.

### DISCUSSION THREAD

**Discussion by anuragnandi** [remix-run/remix#8570] [Q&A] [answered]
After going through a lot of hurdles I managed to put on the ads.txt file, robots.txt and dynamic sitemap.xml to my site and they all work flawlessly. 

A while back my website was approved by google AdSense for monetization, all I need to do was add a link tag and few scripts tags to my website to get the ads showing.  Adding links was very when using remix, the issue comes when I need to load external async scripts. 

All I need to add was the script below, but I get a lot of hydration errors: 

```
  <script
            async
            src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-xxxxxxxxxxxxxxx"
            crossOrigin="anonymous"
          ></script>
```

**The problem:**
1) The ads do no show, the script has been injected into my app when I see the console, but it does not show any ads. I checked and everything is right on google side, its something wrong when the script is loading. There is something wrong with way I am loading the script.
2) google also provides 2 optional script + code for the adBlock detection, I cannot put that into remix at all because the js inside the script is not parsed properly, the project doesn't even compile. 

**What I tried**
dangerouslySetInnerHTML: made no difference
adding the script after mount using the useEffect: made no difference
using defer instead of async: made no difference

**My code**

```
export const Head = () => {
  return (
    <>
      <meta charSet="utf-8" />
      <meta name="viewport" content="width=device-width,initial-scale=1" />
      <meta name="google-adsense-account" content="ca-pub-xxxxxxxxxx" />
      <Meta />
      <Links />
      <ColorSchemeScript />
    </>
  );
};
const GoogleScript = () => {
  const [isMounted, setIsMounted] = React.useState(false);
  React.useEffect(() => {
    setIsMounted(true);
  }, []);
  return (
    <>
      {isMounted && (
        <>
          <script
            async
            src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-xxxxxxxxxx"
            crossOrigin="anonymous"
          ></script>
        </>
      )}
    </>
  );
};

export default function App() {
  const { ENV } = useLoaderData();
  return (
    <html lang="en">
      <head>
        <Head />
        <GoogleScript />
      </head>
      <body itemScope itemType="http://schema.org/WebPage">
          <HeaderMegaMenu />
          <Container
            style={{
              minHeight: "40vh",
            }}
          >
            <ColorSchemeToggle />
            <ScrollRestoration />
            <script
              dangerouslySetInnerHTML={{
                __html: `window.ENV = ${JSON.stringify(ENV)}`,
              }}
            />
            <Scripts />
            <LiveReload />
          </Container>
          <FooterSocial />
      </body>
    </html>
  );
}

```

update:
```
export const AdsScript = () => {
  const [loaded, setLoaded] = useState(false);
  useEffect(() => {
    setLoaded(true);
  }, []);

  if (loaded) {
    const headTag = document.querySelector("head");
    const scriptTag = document.createElement("script");

    scriptTag.setAttribute(
      "src",
      "https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2669682704362378"
    );
    headTag.appendChild(scriptTag);
    headTag.setAttribute("crossorigin", "anonymous");
    headTag.setAttribute("async", "true");
  }
  return null;
};

```

when I add the script in the above way. I can see the ads. But I need to reload the page every time. The first time the page loads, the script is not being executed.

**Comment by cohenerickson** [comment]
Running into a similar issue myself, some input by the devs would be appreciated.

**Comment by kiliman** [comment]
It seems to me that all you need to do is add your script tag inside the `<head>` on the _root_ route. There is no client interactivity, so you shouldn't have to worry about mounting, etc.

Do you have a sample repo that I can check out?

```ts
// root.tsx

return (
  <html>
    <head>
      <script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-1234567890123456" crossorigin="anonymous"></script>
      ...
    </head>
    ...
  </html>
)
```

**Reply by anuragnandi** [reply]
Hi Kiliman! 
The repo is private but I will try to produce a sample. Meanwhile here are few more details: 

1) The above script tag does have client interactivity, It allows google to basically mount HTML elements to DOM and display ads, all over the site.
2) After the page opens first time I see that the script tag is present when I put the tag in the Head component, But no ads are mounted. When I quickly refresh or hit the link again via Menu or Breadcrumbs, the script is executing and I see that the ads are mounted. 

any suggestions?

**Comment by cohenerickson** [comment]
I found an article that explains fairly well the issue that I am facing (but with NextJS), it appears to happen when the adsense script is loaded even without any ad elements, specifically on the Brave browser.

```tsx
export default function App() {
  return (
    <html lang="en">
      <head>
        <script
          async
          src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-1234567890123456"
          crossOrigin="anonymous"
        ></script>
      </head>
      <body>
        {/* <Outlet /> */}
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```

https://deniapps.com/blog/unhandled-runtime-error-hydration-and-google-ads-in-nextjs

**Comment by pie0902** [comment]
Hi, I'm having the same problem. Have you found a solution?

**Comment by KienTPhan** [comment, marked as answer]
After manually adding the script in a useEffect, I use setTimeout before pushing the ad and this seems to work for me. This is inside a component I made to hold an AdUnit.

```ts
const handleAds = () => {
    if (!document.getElementById("adsbygoogleaftermount")) {
      const script = document.createElement("script");
      script.id = "adsbygoogleaftermount";
      script.type = "text/javascript";
      script.async = true;
      script.src =
        "https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-xxxxxxxxxxxxxxx";
      script.crossOrigin = "anonymous";
      document.head.appendChild(script);
    }

    setTimeout(() => {
      // @ts-expect-error adsbygoogle exists
      if (!window.adsbygoogle || initializedRef.current) return; // NOTE: only need to push once

      // @ts-expect-error adsbygoogle exists
      (adsbygoogle = window.adsbygoogle || []).push({});
      initializedRef.current = true;
    }, 200);
  };

  useEffect(() => {
    handleAds();
  }, []);
```

**Reply by anuragnandi** [reply]
works like a charm :) 

just wish there was a proper solution or suggestions form the remix team instead of workarounds..
